core/src/main/clojure/xtdb/expression/temporal.clj

```diff
@@ -902,8 +903,13 @@
    :->call-code (fn [[_ x]]
                   (-> `(-> ~(ts->ldt x ts-unit)
                            ~(case field
+                              "MILLENNIUM" `(-> (.truncatedTo ChronoUnit/DAYS) (.with (MilleniumAdjuster.)) (.withDayOfYear 1))
```

Comment:
just a gut feel, but these seem relatively imperative, especially once there's no clean `truncatedTo` call. maybe more declarative to re-create the LDT? e.g. `(LocalDateTime/of (LocalDate/of (truncate-year year 1000) 1 1) LocalTime/MIDNIGHT)`?

perhaps let's pair, I might be talking BS

Reply:
separately, we can likely avoid the adjuster creation for every row by having those as static final fields instead - this could well be hot-loop code here

Reply:
or, if we did want to do it through the adjusters, we could do the whole adjustment there instead - which would cut down on the amount of generated code